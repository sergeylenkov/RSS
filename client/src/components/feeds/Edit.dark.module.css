.icon {
	composes: icon from './Edit.module.css';
	stroke: #a5a5a5;
	fill: #a5a5a5;
}