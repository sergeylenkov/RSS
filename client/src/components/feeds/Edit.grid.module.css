.container {
  margin: auto 0 auto auto;
}