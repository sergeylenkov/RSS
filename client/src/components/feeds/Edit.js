import React from 'react';
import { Icon, Icons } from '../Icon.js';

import styles from './Edit.module.css';

export class FeedEdit extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            isFormVisible: false
        }

        this.linkFieldRef = React.createRef();
    }

    render() {
        let className = styles.container;

        if (this.props.isDisabled) {
            className += ` ${styles.disabled}`;
        }

        if (this.props.isScrolled) {
            className += ` ${styles.scrolled}`;
        }

        if (this.state.isFormVisible) {
            return (
                <div className={className}>
                    <div className={styles.form}>
                        <input className={styles.linkField} placeholder="ссылка на канал" type="text" ref={this.linkFieldRef} />
                        <button className={styles.formButton} onClick={() => this.onAdd()}><div className={styles.icon}><Icon svg={Icons.add}/></div></button>
                        <button className={styles.formButton} onClick={() => this.onCloseForm()}><div className={styles.icon}><Icon svg={Icons.close}/></div></button>
                    </div>
                </div>
            );
        } else {
            let iconClassName = styles.icon;

            if (this.props.isEdited) {
                iconClassName += ` ${styles.edited}`;
            }

            return (
                <div className={className}>
                    <button className={styles.button} onClick={() => this.onShowForm()}><div className={styles.icon}><Icon svg={Icons.add}/></div></button>
                    <button className={styles.button} onClick={() => this.props.onEdit()}><div className={iconClassName}><Icon svg={Icons.edit}/></div></button>
                </div>
            );
        }
    }

    onAdd() {        
        const value = this.linkFieldRef.current.value;

        if (value.length > 0) {
            this.props.onAdd(value);
        }
    }

    onShowForm() {
        this.setState({
            isFormVisible: true
        });
    }

    onCloseForm() {
        this.setState({
            isFormVisible: false
        });
    }
}