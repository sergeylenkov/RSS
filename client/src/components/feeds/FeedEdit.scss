.feed-edit {
  display: flex;
  height: 30px;
  margin: 30px 0 0 0;
  padding: 0;
  opacity: 1;
  transition: opacity 0.2s ease-in-out;

  &_disabled {
    opacity: 0.6;
    pointer-events: none;
  }

  &_scrolled {
    opacity: 0;
    pointer-events: none;
  }

  &__button {
    display: flex;
    margin: auto 10px auto 0;
    padding: 0;
    cursor: pointer;
  }

  &__icon {
    width: 16px;
    height: 16px;
    fill: #000000;
    stroke: #000000;
    stroke-width: 2px;

    &_editing {
      stroke: #2f74f1;
    }

    svg {
      width: 100%;
      height: 100%;
    }
  }
}

.feed-edit-form {
  display: flex;
  width: 100%;


  &__button {
    display: flex;
    margin: auto 0 auto 10px;
    padding: 0;
    cursor: pointer;
  }

  &__field {
    flex-grow: 1;
    margin: auto 0;
    font-size: 14px;
  }
}

/*
* grid
*/

.grid .feed-edit {
  margin: auto 0 auto auto;
}

/*
* theme
*/

.dark .feed-edit {
  &__icon {
    fill: #a5a5a5;
    stroke: #a5a5a5;
  }
}

@media (max-width: 768px) {
  .feed-edit {
    margin: 0;
  }
}
