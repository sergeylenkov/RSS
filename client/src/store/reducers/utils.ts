import { Entry, Feed } from '../../data';
import { State } from './index';

const feedsDict: { [key: string]: Feed } = {};

export function updateFeeds(state: State, feeds: Feed[]): State {
  feeds.forEach((feed: Feed) => {
    feedsDict[feed.id] = feed;

    const iconId = `icon_${feed.id}`;

    feed.icon = localStorage.getItem(iconId) || feed.iconUrl;

    if (!localStorage.getItem(iconId)) {
      toDataURL(feed.iconUrl).then(data => {
        localStorage.setItem(iconId, data);
      });    
    }
  });

  return {
    ...state,
    isUpdating: false,
    isInitialized: true,
    feeds: [...feeds],
  };
}

export function updateFeed(state: State, id: number, newFeed: Feed): State {
  const feeds = state.feeds.map((feed: Feed) => {
    return feed.id === id ? { ...feed, ...newFeed } : feed;
  });

  return {
    ...state,
    feeds: feeds,
  };
}

export function addFeed(state: State, feed: Feed): State {
  const feeds = [...state.feeds];

  feeds.push(feed);

  return {
    ...state,
    feeds: feeds,
  };
}

export function removeFeed(state: State, id: number): State {
  const feeds = state.feeds.filter((feed: Feed) => {
    return feed.id !== id;
  });

  return {
    ...state,
    feeds: feeds,
  };
}

export function selectFeed(state: State, id: number): State {
  const selectedFeeds = [...state.selectedFeeds];

  const index = selectedFeeds.indexOf(id);

  if (index === -1) {
    selectedFeeds.push(id);
  } else {
    selectedFeeds.splice(index, 1);
  }

  const entries = filterEntries(selectedFeeds, state.allEntries);

  return {
    ...state,
    selectedFeeds: selectedFeeds,
    entries: [...entries],
  };
}

export function updateUnviewedCount(state: State): State {
  const feeds = [...state.feeds];
  let totalCount = 0;

  feeds.forEach((feed: Feed) => {
    const count = state.entries.filter((entry: Entry) => {
      return entry.feedId === feed.id && !entry.isViewed;
    }).length;

    feed.count = count;
    totalCount = totalCount + count;
  });

  return {
    ...state,
    feeds: feeds,
    unviewedCount: totalCount,
  };
}

export function updateViewed(state: State, ids: number[]): State {
  let entries = state.entries.map((entry: Entry) => {
    if (ids.indexOf(entry.id) !== -1) {
      return { ...entry, isViewed: true };
    }

    return entry;
  });

  entries = filterEntries(state.selectedFeeds, entries);

  return {
    ...state,
    entries: entries,
  };
}

export function updateFavorite(
  state: State,
  id: number,
  isFavorite: boolean
): State {
  let entries = state.entries.map((entry: Entry) => {
    return entry.id === id ? { ...entry, isFavorite: isFavorite } : entry;
  });

  entries = filterEntries(state.selectedFeeds, entries);

  return {
    ...state,
    entries: entries,
  };
}

export function updateRead(state: State, id: number, isRead: boolean): State {
  let entries = state.entries.map((entry: Entry) => {
    return entry.id === id ? { ...entry, isRead: isRead } : entry;
  });

  entries = filterEntries(state.selectedFeeds, entries);

  return {
    ...state,
    entries: entries,
  };
}

export function updateEntries(state: State, entries: Entry[]): State {
  updateFeedsInEntries(entries);

  const filteredEntries = filterEntries(state.selectedFeeds, entries);

  return {
    ...state,
    isUpdating: false,
    entriesCount: entries.length,
    entries: [...filteredEntries],
    allEntries: [...entries],
  };
}

export function updateEntriesCount(state: State): State {
  const feeds = [...state.feeds];

  feeds.forEach((feed: Feed) => {
    feed.count = state.entries.filter((entry: Entry) => {
      return entry.feedId === feed.id;
    }).length;
  });

  return {
    ...state,
    feeds: feeds,
  };
}

function filterEntries(selectedFeeds: number[], allEntries: Entry[]): Entry[] {
  let entries: Entry[] = [];

  if (selectedFeeds.length > 0) {
    entries = allEntries.filter((entry: Entry) => {
      return selectedFeeds.includes(entry.feedId);
    });
  } else {
    entries = allEntries;
  }

  return entries;
}

function getFeedById(id: number): Feed {
  return feedsDict[id];
}

function updateFeedsInEntries(entries: Entry[]): void {
  entries.forEach((entry) => {
    entry.feed = getFeedById(entry.feedId);
  });
}

function toDataURL(url: string): Promise<string> {
  return new Promise((resolve, reject) => {
    fetch(url)
    .then(response => response.blob())
    .then(blob => {
      const reader = new FileReader()

      reader.onloadend = () => resolve(reader.result as string)
      reader.onerror = reject
      reader.readAsDataURL(blob)
    });
  });  
}