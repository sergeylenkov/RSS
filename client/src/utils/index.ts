export const debounce = <F extends (...args: unknown[]) => unknown>(
  func: F,
  waitFor: number
) => {
  let timeout = 0;

  const debounced = (...args: unknown[]) => {
    clearTimeout(timeout);
    timeout = setTimeout(() => func(...args), waitFor) as unknown as number;
  };

  return debounced as unknown as (...args: Parameters<F>) => ReturnType<F>;
};

export function isSystemDarkTheme(): boolean {
  return (
    window.matchMedia &&
    window.matchMedia('(prefers-color-scheme: dark)').matches
  );
}
