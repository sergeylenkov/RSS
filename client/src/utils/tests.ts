export const action = (): void => { return };