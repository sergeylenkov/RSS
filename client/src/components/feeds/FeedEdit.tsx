import React, { useRef, useState } from 'react';
import { Bem } from '../../utils/bem';
import { CheckmarkIcon, CloseIcon, AddIcon, EditIcon } from '../Icons';

import './FeedEdit.scss';

const block = new Bem('feed-edit');
const fromBlock = new Bem('feed-edit-form');

interface FeedEditProps {
  isEditing: boolean;
  isDisabled?: boolean;
  isScrolled?: boolean;
  onEdit: () => void;
  onAdd: (link: string) => void;
}

export function FeedEdit({
  isDisabled,
  isScrolled,
  isEditing,
  onEdit,
  onAdd,
}: FeedEditProps): JSX.Element {
  const [isFormVisible, setFormVisible] = useState(false);
  const linkFieldRef = useRef<HTMLInputElement>(null);

  const onAddClick = () => {
    if (linkFieldRef && linkFieldRef.current) {
      const value = linkFieldRef.current.value;

      if (value.length > 0) {
        onAdd(value);
      }
    }
  };

  const blockClass = block
    .addModifier(isDisabled ? 'disabled' : '')
    .addModifier(isScrolled ? 'scrolled' : '')
    .toString();
  const iconClass = block.getElement('icon').toString();
  const editIconClass = block
    .getElement('icon')
    .addModifier(isEditing ? 'editing' : '')
    .toString();
  const buttonClass = block.getElement('button').toString();
  const formButtonClass = fromBlock.getElement('button').toString();

  if (isFormVisible) {
    return (
      <div className={blockClass}>
        <div className={fromBlock.toString()}>
          <input
            ref={linkFieldRef}
            className={fromBlock.getElement('field').toString()}
            placeholder="ссылка на канал"
            type="text"
          />
          <button className={formButtonClass} onClick={onAddClick}>
            <div className={iconClass}>
              <CheckmarkIcon />
            </div>
          </button>
          <button
            className={formButtonClass}
            onClick={() => setFormVisible(false)}
          >
            <div className={iconClass}>
              <CloseIcon />
            </div>
          </button>
        </div>
      </div>
    );
  } else {
    return (
      <div className={blockClass}>
        <button className={buttonClass} onClick={() => setFormVisible(true)}>
          <div className={iconClass}>
            <AddIcon />
          </div>
        </button>
        <button className={buttonClass} onClick={onEdit}>
          <div className={editIconClass}>
            <EditIcon />
          </div>
        </button>
      </div>
    );
  }
}
