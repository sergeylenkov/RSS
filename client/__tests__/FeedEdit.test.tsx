import React from 'react';
import renderer from 'react-test-renderer';
import { FeedEdit } from '../src/components/feeds/FeedEdit';
import { action } from '../src/utils/tests';

test('FeedEdit isEditing = true', () => {
  const render = renderer.create(
    <FeedEdit
      isEditing
      isDisabled={false}
      isScrolled={false}
      onAdd={action}
      onEdit={action}
    />
  );
  expect(render).toMatchSnapshot();
});

test('FeedEdit isDisabled = true', () => {
  const render = renderer.create(
    <FeedEdit
      isEditing={false}
      isDisabled
      isScrolled={false}
      onAdd={action}
      onEdit={action}
    />
  );
  expect(render).toMatchSnapshot();
});

test('FeedEdit isScrolled = true', () => {
  const render = renderer.create(
    <FeedEdit
      isEditing={false}
      isDisabled={false}
      isScrolled
      onAdd={action}
      onEdit={action}
    />
  );
  expect(render).toMatchSnapshot();
});
