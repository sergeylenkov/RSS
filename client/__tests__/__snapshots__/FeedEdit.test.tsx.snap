// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedEdit isDisabled = true 1`] = `
<div
  className="feed-edit feed-edit_disabled"
>
  <button
    className="feed-edit__button"
    onClick={[Function]}
  >
    <div
      className="feed-edit__icon"
    >
      <svg
        viewBox="0 0 24 24"
      >
        <g
          fill="none"
        >
          <line
            x1="12"
            x2="12"
            y1="5"
            y2="19"
          />
          <line
            x1="5"
            x2="19"
            y1="12"
            y2="12"
          />
        </g>
      </svg>
    </div>
  </button>
  <button
    className="feed-edit__button"
    onClick={[Function]}
  >
    <div
      className="feed-edit__icon"
    >
      <svg
        viewBox="0 0 24 24"
      >
        <polygon
          fill="none"
          points="16 3 21 8 8 21 3 21 3 16 16 3"
          strokeLinecap="round"
          strokeLinejoin="round"
        />
      </svg>
    </div>
  </button>
</div>
`;

exports[`FeedEdit isEditing = true 1`] = `
<div
  className="feed-edit"
>
  <button
    className="feed-edit__button"
    onClick={[Function]}
  >
    <div
      className="feed-edit__icon"
    >
      <svg
        viewBox="0 0 24 24"
      >
        <g
          fill="none"
        >
          <line
            x1="12"
            x2="12"
            y1="5"
            y2="19"
          />
          <line
            x1="5"
            x2="19"
            y1="12"
            y2="12"
          />
        </g>
      </svg>
    </div>
  </button>
  <button
    className="feed-edit__button"
    onClick={[Function]}
  >
    <div
      className="feed-edit__icon feed-edit__icon_editing"
    >
      <svg
        viewBox="0 0 24 24"
      >
        <polygon
          fill="none"
          points="16 3 21 8 8 21 3 21 3 16 16 3"
          strokeLinecap="round"
          strokeLinejoin="round"
        />
      </svg>
    </div>
  </button>
</div>
`;

exports[`FeedEdit isScrolled = true 1`] = `
<div
  className="feed-edit feed-edit_scrolled"
>
  <button
    className="feed-edit__button"
    onClick={[Function]}
  >
    <div
      className="feed-edit__icon"
    >
      <svg
        viewBox="0 0 24 24"
      >
        <g
          fill="none"
        >
          <line
            x1="12"
            x2="12"
            y1="5"
            y2="19"
          />
          <line
            x1="5"
            x2="19"
            y1="12"
            y2="12"
          />
        </g>
      </svg>
    </div>
  </button>
  <button
    className="feed-edit__button"
    onClick={[Function]}
  >
    <div
      className="feed-edit__icon"
    >
      <svg
        viewBox="0 0 24 24"
      >
        <polygon
          fill="none"
          points="16 3 21 8 8 21 3 21 3 16 16 3"
          strokeLinecap="round"
          strokeLinejoin="round"
        />
      </svg>
    </div>
  </button>
</div>
`;
